### PR TITLE
S08 judging

### PR DESCRIPTION
**Pull request by Sarosim**
## Description

This PR includes the rendering of the judging page according to the ‘Judging Page’ wireframe, as well as handling the score submission and saving it to the model.

## Pull request type

S08 As Staff, I would like to Judge each submission.

## Related Issue

none existing, one question to be decided:
The Project Score Categories model doesn't have description field for the score categories, so the wireframes couldn't be fully followed. I have included screenshots with and without descriptions, if we want to have the descriptions, the model needs to be adjusted.

## Configuration instructions

needs to have Hackathon event with finish date in the past, participating team in the event, valid user as judge for the event, project assigned to the team and scores to be non existent for the hackathon and the team by the judge. Also, project score categories have to be present in the respective model.

## Testing

Four verifications are included:
-	whether user is judge for the event
-	if event is ready to be judged (finished)
-	whether the selected team belongs to the selected event
-	if the judge has already scored the requested team's Project (Issue #37 says: “Cannot edit scores once submitted”)
all four tested.

Scores cannot be submitted without valid score for each and every project score categories
Responsiveness tested in Chrome Dev Tools screen shots included

## Screenshots

image

image

image

## Additional Information

none

## Does this introduce a breaking change

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by Sarosim** [comment]
Pulled the latest PRs from upstream, updated this branch and fixed merge conflicts.

File: hackathon/views.py [hunk 1]
```diff
@@ -17,6 +17,81 @@ class HackathonListView(ListView):
...
+    for judge in event.values('judges'):
+        if judge['judges'] == request.user.id:
+            user_is_judge = True
+    if not user_is_judge:
+        messages.error(request, "You are not a judge for that event!")
+        template = 'home/index.html'
+        return render(request, template)
+
+    # verify if event is ready to be judged (finished)
+    finish = event.values('end_date')[0]['end_date']
```
Comment (stefdworschak) [inline hackathon/views.py:41]: Not sure this is the right trigger. If the event is finished, it is over and done with I would think. Maybe instead of that we can add another status to the hackathon called `judging` and add that.

File: hackathon/views.py [hunk 2]
```diff
@@ -17,6 +17,81 @@ class HackathonListView(ListView):
...
+    now = timezone.now()
+    if finish > now:
+        messages.error(request, f"The event has not finished yet, check back after {finish}!")
+        template = 'home/index.html'
+        return render(request, template)
+
+    # verify that the selected team belongs to the selected event
+    team_belongs_to_event = False
+    for team in event.values('teams'):
```
Comment (stefdworschak) [inline hackathon/views.py:50]: Again this could probably be simplified.

File: hackathon/views.py [hunk 3]
```diff
@@ -17,6 +17,81 @@ class HackathonListView(ListView):
...
+        if team['teams'] == int(team_id):
+            team_belongs_to_event = True
+    if not team_belongs_to_event:
+        messages.error(request, f"Nice try! That team is not part of the event...")
+        template = 'home/index.html'
+        return render(request, template)
+
+    # check if the judge has already scored the requested team's Project
+    the_event = get_object_or_404(Hackathon, pk=hack_id)
```
Comment (stefdworschak) [inline hackathon/views.py:59]: You already have this defined further up, so I would reuse that instance.

**Review by stefdworschak — CHANGES_REQUESTED** [review]
@Sarosim the PR looks good. There are a few comments for you. I think the main thing to look at is using an extra `judging` status in the hackathon instead of the `end_date` to check if judging should be allowed.

Also, can you create a follow-up PR to create the page to list all team projects for a hackathon and link from the hackthon to the list of all projects page and then to the scoring page?

**Comment by Sarosim** [comment]
reviewed all the change requests and solved them, have not made migrations but pushed all other changes.

File: hackathon/urls.py [hunk 4]
```diff
@@ -1,9 +1,11 @@
 from django.urls import path
-
+from . import views
```
Comment (stefdworschak) [inline hackathon/urls.py:2]: Could you be more explicit in importing your views?

File: hackathon/views.py [hunk 5]
```diff
@@ -30,6 +30,68 @@ def get_context_data(self, **kwargs):
...
+                judge = request.user,
+                project = get_object_or_404(HackTeam, pk=team_id).project,
+                score = request.POST.get(score_cat_id),
+                hack_project_score_category = score_category,
+            )
+            team_score.save()
+            
+        return redirect("hackathon:hackathon-list")
+
+    selected_team = get_object_or_404(HackTeam, pk=team_id)
```
Comment (stefdworschak) [inline hackathon/views.py:83]: Could you please remove this and the following line as they are no longer needed?

**Review by stefdworschak — APPROVED** [review]
@Sarosim LGTM

Please just fix the two small comments and I will merge this.

File: hackathon/urls.py (hunk 4 above)
Comment (Sarosim) [reply hackathon/urls.py]: In the new django lessons Chris is importing only the views and using views.judging in the path. I thought it was nicer, but prefer importing the function as you suggested 😉

File: hackathon/views.py (hunk 5 above)
Comment (Sarosim) [reply hackathon/views.py]: done

**Comment by Sarosim** [comment]
All done @stefdworschak

File: accounts/fixtures/profiles.json [hunk 6]
```diff
@@ -1,4 +1,26 @@
 [
+    {
```
Comment (stefdworschak) [inline accounts/fixtures/profiles.json:2]: I think this is a duplicate now?

File: scripts/seed.sh [hunk 7]
```diff
@@ -2,7 +2,6 @@ echo "============================"
 echo "Seeding fixtures"
 echo "============================"
 python3 manage.py loaddata organisation
-python3 manage.py loaddata accounts
```
Comment (stefdworschak) [inline scripts/seed.sh:5]: Why did you remove this?

File: hackathon/views.py [hunk 8]
```diff
@@ -1,13 +1,13 @@
 from datetime import datetime
-
 from django.views.generic import ListView
 from django.shortcuts import render, get_object_or_404, redirect
 from django.contrib import messages
 from django.contrib.auth.decorators import login_required
-
-from .models import Hackathon
+from django.utils import timezone
+from .models import Hackathon, HackTeam, HackProject, HackProjectScore, HackProjectScoreCategory
```
Comment (stefdworschak) [inline hackathon/views.py:7]: There should be a blank line between 3rd party and local modules.

**Review by stefdworschak — COMMENTED** [review]
@Sarosim not sure if I overlooked some of these things, but I think you introduced some new chnages here. Could you please make the few additional changes before merging? Thanks

File: hackathon/views.py (hunk 8 above)
Comment (Sarosim) [reply hackathon/views.py:7]: done

**Comment by Sarosim** [comment]
all done @stefdworschak

**Review by stefdworschak — APPROVED** [review]
LGTM
